### PR TITLE
Drop torch attributes at the end of backend conversion.

### PR DESCRIPTION
**Pull request by ScottTodd**
Fixes https://github.com/llvm/torch-mlir/issues/2866

Some backends / downstream projects expect that a "fully converted" program has no remaining ops or attributes from the original dialect(s).

### CODE REVIEW

File: lib/Dialect/TorchConversion/Transforms/BackendTypeConversionPasses.cpp [hunk 1]
```diff
@@ -115,6 +115,15 @@ static void setupFinalization(ConversionTarget &target,
   setupFinalization<OpTy2, OpTys...>(target, patterns, typeConverter);
 }
 
+static void stripTorchAttrs(func::FuncOp func) {
+  SmallVector<NamedAttribute> newAttrs;
+  for (auto attr : func->getDialectAttrs()) {
```
Comment (ScottTodd) [inline lib/Dialect/TorchConversion/Transforms/BackendTypeConversionPasses.cpp:120]: I'm not very familiar with how torch-mlir uses attributes. This could also look through `getAllArgAttrs()` and `getAllResultAttrs()`.

File: lib/Dialect/TorchConversion/Transforms/BackendTypeConversionPasses.cpp [hunk 2]
```diff
@@ -151,6 +160,9 @@ struct FinalizingBackendTypeConversionPass
 
     if (failed(applyFullConversion(func, target, std::move(patterns))))
       signalPassFailure();
+
+    // Drop attributes that are no longer used after conversion out of Torch.
+    stripTorchAttrs(func);
```
Comment (ScottTodd) [inline lib/Dialect/TorchConversion/Transforms/BackendTypeConversionPasses.cpp:165]: Are there users of torch-mlir that would miss these attributes after conversion? I could make this optional (via a separate pass, an option here, etc.)

File: test/Conversion/TorchToLinalg/basic.mlir [hunk 3]
```diff
@@ -30,7 +30,7 @@ func.func @torch.aten.mm$basic(%arg0: !torch.vtensor<[?,?],f32>, %arg1: !torch.v
 // -----
 
 // CHECK-LABEL: func.func @torch.aten.matmul.2d
-func.func @torch.aten.matmul.2d(%arg0: !torch.vtensor<[8,16],f32>, %arg1: !torch.vtensor<[16,8],f32>) -> !torch.vtensor<[8,8],f32> attributes {torch.onnx_meta.ir_version = 7 : si64, torch.onnx_meta.opset_version = 13 : si64, torch.onnx_meta.producer_name = "backend-test", torch.onnx_meta.producer_version = ""} {
+func.func @torch.aten.matmul.2d(%arg0: !torch.vtensor<[8,16],f32>, %arg1: !torch.vtensor<[16,8],f32>) -> !torch.vtensor<[8,8],f32> {
```
Comment (ScottTodd) [inline test/Conversion/TorchToLinalg/basic.mlir:33]: This diff isn't directly related to this change, but I felt like including it (the attributes are irrelevant to the test case). Could revert or move to another PR.

File: lib/Dialect/TorchConversion/Transforms/BackendTypeConversionPasses.cpp (hunk 1 above)
Comment (stellaraccident) [reply lib/Dialect/TorchConversion/Transforms/BackendTypeConversionPasses.cpp:120]: It is relatively expensive to mutate these attributes. Let's make it only pay that cost if there are attributes to change. Can you carry a `bool modified` outside the loop to note if you dropped any and then only setDialectAttrs if modified?

File: lib/Dialect/TorchConversion/Transforms/BackendTypeConversionPasses.cpp (hunk 2 above)
Comment (stellaraccident) [reply lib/Dialect/TorchConversion/Transforms/BackendTypeConversionPasses.cpp:165]: I don't think so. Historically there haven't been any/many torch dialect attributes used. I suspect that we just added some and forgot to make the conversion "full".

File: test/Conversion/TorchToLinalg/basic.mlir (hunk 3 above)
Comment (stellaraccident) [reply test/Conversion/TorchToLinalg/basic.mlir:33]: Makes sense to me. Fine to include.

**Review by stellaraccident — APPROVED** [review]
I'd like to see the conditional modification I note in the comments. Otherwise, lgtm.

File: lib/Dialect/TorchConversion/Transforms/BackendTypeConversionPasses.cpp (hunk 1 above)
Comment (ScottTodd) [reply lib/Dialect/TorchConversion/Transforms/BackendTypeConversionPasses.cpp:121]: Added a modified check and just a comment about how "arg" and "result" attrs could also be stripped if needed.
